src/engine/strat_engine/backstore/setup.rs

```diff
@@ -93,28 +94,26 @@ pub fn get_metadata(
     // Try to read from all available devnodes that could contain most
     // recent metadata. In the event of errors, continue to try until all are
     // exhausted.
-    for &(devnode, ref bda) in bdas
-        .iter()
-        .filter(|&&(_, ref bda)| bda.last_update_time() == Some(most_recent_time))
-    {
-        let poolsave = OpenOptions::new()
-            .read(true)
-            .open(devnode)
-            .ok()
-            .and_then(|mut f| bda.load_state(&mut f).ok())
-            .and_then(|opt| opt)
-            .and_then(|data| serde_json::from_slice(&data).ok());
-
-        if poolsave.is_some() {
-            return Ok(poolsave);
-        }
-    }
-
-    // If no data has yet returned, we have an error. That is, we should have
-    // some metadata, because we have a most recent time, but we failed to
-    // get any.
-    let err_str = "timestamp indicates data was written, but no data successfully read";
-    Err(StratisError::Engine(ErrorEnum::NotFound, err_str.into()))
+    bdas.iter()
+        .filter_map(|&(devnode, ref bda)| {
+            if bda.last_update_time() == Some(most_recent_time) {
+                OpenOptions::new()
+                    .read(true)
+                    .open(devnode)
+                    .ok()
+                    .and_then(|mut f| bda.load_state(&mut f).ok())
```

Comment:
Err, could you do one more thing? Apparently this construction has troubled me since two years ago, PR#444. Change these two lines to the equivalent single line:

```
.and_then(|mut f| bda.load_state(&mut f).unwrap_or(None))
```